flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/co/CoBroadcastWithKeyedOperatorTest.java

```diff
@@ -556,9 +559,27 @@ public void testScaleUp() throws Exception {
                                 3,
                                 2,
                                 operatorSubtaskState3)) {
-            testHarness1.processElement1(new StreamRecord<>("trigger"));
-            testHarness2.processElement1(new StreamRecord<>("trigger"));
-            testHarness3.processElement1(new StreamRecord<>("trigger"));
+
+            // Since there is a keyed operator, we should follow the key partition rules.
+            Map<TwoInputStreamOperatorTestHarness<String, Integer, String>, KeyGroupRange>
+                    keyGroupPartition = new HashMap<>();
+            keyGroupPartition.put(testHarness1, KeyGroupRange.of(0, 3));
+            keyGroupPartition.put(testHarness2, KeyGroupRange.of(4, 6));
+            keyGroupPartition.put(testHarness3, KeyGroupRange.of(7, 9));
+            while (!keyGroupPartition.isEmpty()) {
+                String triggerKey = String.valueOf(ThreadLocalRandom.current().nextLong());
+                for (Map.Entry<
+                                TwoInputStreamOperatorTestHarness<String, Integer, String>,
+                                KeyGroupRange>
+                        entry : keyGroupPartition.entrySet()) {
+                    if (entry.getValue()
+                            .contains(KeyGroupRangeAssignment.assignToKeyGroup(triggerKey, 10))) {
+                        entry.getKey().processElement1(new StreamRecord<>(triggerKey));
+                        keyGroupPartition.remove(entry.getKey());
+                        break;
+                    }
+                }
+            }
```

Comment:
This part doesn't seem obvious to me; and mapping from harnesses to key ranges can break.
How about this:
```
            for (TwoInputStreamOperatorTestHarness<String, Integer, String> harness :
                    Arrays.asList(testHarness1, testHarness2, testHarness3)) {
                int subtask = harness.getEnvironment().getTaskInfo().getIndexOfThisSubtask();
                // find the right input element for this subtask
                int element = 0;
                while (assignKeyToParallelOperator(Integer.toString(element), 10, 3) != subtask) {
                    element++;
                }
                harness.processElement1(new StreamRecord<>(Integer.toString(element)));
            }
```
?

ditto: `testScaleDown` - extract a function?

Reply:
Sure, I extracted a function for finding a valid key for a subtask.

Reply:
Thanks!